utils/filesystem.c

```diff
@@ -0,0 +1,414 @@
+    // create the last directory, this happend when the given path
+    // doesn't finish with '/'
+    if (created < strlen(path)) {
+        if (mkdir(path, mode) != 0 && errno != EEXIST)
+            return -1;
+    }
+
+    return 0;
+}
+
+void pv_filesystem_path_join(char *buf, int size, ...)
+{
+    char fmt[PATH_MAX] = { 0 };
+
+    for (int i = 0; i < size; ++i) {
+        fmt[i * 3 + 0] = '%';
+        fmt[i * 3 + 1] = 's';
+        fmt[i * 3 + 2] = '/';
+    }
+
+    va_list list;
+    va_start(list, size);
+
+    vsnprintf(buf, PATH_MAX, fmt, list);
+    buf[strnlen(buf, PATH_MAX) - 1] = '\0';
+
+    va_end(list);
+}
+
+int pv_filesystem_path_remove(const char *path, bool recursive)
+{
+    if (!recursive) {
+        return remove(path);
+    }
+
+    struct dirent **arr = NULL;
+    int n = scandir(path, &arr, NULL, alphasort);
+
+    for (int i = 0; i < n; ++i) {
+        // discard . and .. from scandir
+        if (!strcmp(arr[i]->d_name, ".") || !strcmp(arr[i]->d_name, ".."))
+            goto free_dir;
+
+        char new_path[PATH_MAX] = { 0 };
+        pv_filesystem_path_join(new_path, 2, path, arr[i]->d_name);
+
+        if (arr[i]->d_type == DT_DIR)
+            pv_filesystem_path_remove(new_path, true);
+        else
+            pv_filesystem_path_remove(new_path, false);
+
+free_dir:
+        free(arr[i]);
+    }
+    int ret = remove(path);
+    free(arr);
+
+    return ret;
+}
+
+int pv_filesystem_path_rename(const char *src_path, const char *dst_path)
+{
+    pv_filesystem_path_sync(src_path);
+
+    int ret = rename(src_path, dst_path);
+    if (ret < 0)
+        return ret;
+
+    pv_filesystem_path_sync(dst_path);
+    return 0;
+}
+
+int pv_filesystem_file_tmp(char *tmp, const char *fname)
+{
+    if (!fname)
+        return -1;
+
+    size_t size = strnlen(fname, PATH_MAX) + 5;
+
+    if (size > PATH_MAX) {
+        errno = ENAMETOOLONG;
+        return -1;
+    }
+
+    snprintf(tmp, size, "%s.tmp", fname);
+    return 0;
+}
+
+int pv_filesystem_file_save(const char *fname, const char *data, mode_t mode)
+{
+    char tmp[PATH_MAX] = { 0 };
+    if (pv_filesystem_file_tmp(tmp, fname) != 0)
+        return -1;
+
+    int ret = -1;
+    int fd = open(tmp, O_CREAT | O_WRONLY | O_TRUNC | O_SYNC, mode);
+    if (fd < 0)
+        goto out;
+
+    if (write(fd, data, strlen(data)) < 0)
+        goto out;
+
+    close_fd(&fd);
+
+    ret = pv_filesystem_path_rename(tmp, fname);
+
+out:
+    if (fd > 0)
+        close_fd(&fd);
+
+    ret = pv_filesystem_path_remove(tmp, false);
+    pv_filesystem_path_sync(tmp);
+
+    return ret;
+}
+
+ssize_t pv_filesystem_file_copy_from_fd(int src, int dst, bool close_src)
+{
+    lseek(src, 0, SEEK_SET);
+    lseek(dst, 0, SEEK_SET);
+
+    char buf[4096] = { 0 };
+    ssize_t read_bytes = 0;
+    ssize_t write_bytes = 0;
+
+    while (read_bytes = read(src, buf, 4096), read_bytes > 0)
+        write_bytes += write(dst, buf, read_bytes);
+
+    if (close_src)
+        close_fd(&src);
+
```

Comment:
we have to sync too. In general, we sync because we can have some problems with files that should or shouldn't exist in low spec devices with slow disk throughput, so check if we need more in the rest of the code

Reply:
Ok, I will kept an eye on it

Reply:
Here cannot sync, because we haven't a path, so maybe it's a good idea only to warn about this issue.
In `/proc/...` but it feels a bit hacky and not very failure proof